### ISSUE TITLE
Failed to allocate memory for memtx_tuple if tuples size is increasing

### ISSUE THREAD

**Issue by UnDeleteRU** [tarantool/tarantool#3633]
We using space to store some tuples which can grow with time. After some point some tuples can not be saved because of error 'Failed to allocate 138301 bytes in slab allocator for memtx_tuple'. Even if we have enough free memory
```
localhost:3301> box.slab.info()
---
- items_size: 519657728
  items_used_ratio: 28.10%
  quota_size: 536870912
  quota_used_ratio: 100.00%
  arena_used_ratio: 29.1%
  items_used: 146033064
  quota_used: 536870912
  arena_size: 536870912
  arena_used: 156273064
...

localhost:3301> box.info.version
---
- 1.9.0-70-g886be65
...
```

We reproduce this behavior with this script
```lua
s = box.schema.space.create('test')
s:create_index('primary', {
    if_not_exists = true,
    type = 'tree',
    parts = {1, 'integer'}
})
d = require 'digest'
box.slab.info()

size = 500 * 1024
str = d.urandom(size)
for i = 1,1024 do s:insert({i, str}) end
box.slab.info()

s:truncate()
collectgarbage('collect')

for j = 9,40 do
    size = j * 8 * 1024
    str = d.urandom(size)
    for i = 1,1024 do s:insert({i, str}) end
    print(size, s:len(), s:bsize(), box.slab.info().quota_used_ratio, box.slab.info().items_used_ratio, box.slab.info().items_used_ratio)
    for i = 1,1024 do s:delete(i) end
    collectgarbage('collect')
end

s:len(), s:bsize()
box.slab.info()
```
Output on local machine 
```
localhost:3301> s = box.schema.space.create('test')
---
...

localhost:3301> s:create_index('primary', {
              >     if_not_exists = true,
              >     type = 'tree',
              >     parts = {1, 'integer'}
              > })
---
- unique: true
  parts:
  - type: integer
    is_nullable: false
    fieldno: 1
  id: 0
  space_id: 520
  name: primary
  type: TREE
...

localhost:3301> d = require 'digest'
---
...

localhost:3301> box.slab.info()
---
- items_size: 423904
  items_used_ratio: 2.07%
  quota_size: 536870912
  quota_used_ratio: 6.25%
  arena_used_ratio: 3.9%
  items_used: 8776
  quota_used: 33554432
  arena_size: 33554432
  arena_used: 1303112
...

localhost:3301> 
---
...

localhost:3301> size = 500 * 1024
---
...

localhost:3301> str = d.urandom(size)
---
...

localhost:3301> for i = 1,1024 do s:insert({i, str}) end
2018-08-21 18:41:59.669 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 563: 1.045 sec
---
- error: Failed to allocate 512025 bytes in slab allocator for memtx_tuple
...

localhost:3301> box.slab.info()
---
- items_size: 503737264
  items_used_ratio: 98.41%
  quota_size: 536870912
  quota_used_ratio: 100.00%
  arena_used_ratio: 92.6%
  items_used: 495706216
  quota_used: 536870912
  arena_size: 536870912
  arena_used: 497049704
...

localhost:3301> 
---
...

localhost:3301> s:truncate()
---
...

localhost:3301> collectgarbage('collect')
---
- 0
...

localhost:3301> 
---
...

localhost:3301> for j = 9,40 do
              >     size = j * 8 * 1024
              >     str = d.urandom(size)
              >     for i = 1,1024 do s:insert({i, str}) end
              >     print(size, s:len(), s:bsize(), box.slab.info().quota_used_ratio, box.slab.info().items_used_ratio, box.slab.info().items_used_ratio)
              >     for i = 1,1024 do s:delete(i) end
              >     collectgarbage('collect')
              > end
2018-08-21 18:42:02.029 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 1797: 1.031 sec
73728	1024	75506306ULL	100.00%	76.64%	76.64%
81920	1024	83894914ULL	100.00%	72.48%	72.48%
90112	1024	92283522ULL	100.00%	69.96%	69.96%
2018-08-21 18:42:04.385 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 7902: 1.032 sec
98304	1024	100672130ULL	100.00%	61.74%	61.74%
106496	1024	109060738ULL	100.00%	61.90%	61.90%
114688	1024	117449346ULL	100.00%	55.02%	55.02%
2018-08-21 18:42:06.654 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 13408: 1.043 sec
122880	1024	125837954ULL	100.00%	53.65%	53.65%
131072	1024	134226562ULL	100.00%	51.77%	51.77%
2018-08-21 18:42:08.896 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 17545: 1.024 sec
139264	1024	142615170ULL	100.00%	51.17%	51.17%
147456	1024	151003778ULL	100.00%	48.08%	48.08%
2018-08-21 18:42:11.177 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 21458: 1.034 sec
155648	1024	159392386ULL	100.00%	47.96%	47.96%
2018-08-21 18:42:13.362 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 24171: 1.021 sec
163840	1024	167780994ULL	100.00%	45.79%	45.79%
172032	1024	176169602ULL	100.00%	45.99%	45.99%
2018-08-21 18:42:15.641 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 27786: 1.030 sec
180224	1024	184558210ULL	100.00%	44.43%	44.43%
2018-08-21 18:42:17.890 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 30271: 1.025 sec
188416	1024	192946818ULL	100.00%	44.86%	44.86%
2018-08-21 18:42:20.091 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 32677: 1.023 sec
196608	1024	201335426ULL	100.00%	43.74%	43.74%
204800	1024	209724034ULL	100.00%	44.35%	44.35%
2018-08-21 18:42:22.318 [29318] main/108/main txn.c:277 W> too long WAL write: 1 rows at LSN 36027: 1.044 sec
---
- error: Failed to allocate 213017 bytes in slab allocator for memtx_tuple
...

localhost:3301> 
---
...

localhost:3301> s:len(), s:bsize()
---
- 975
- 207675593
...

localhost:3301> box.slab.info()
---
- items_size: 503753440
  items_used_ratio: 42.86%
  quota_size: 536870912
  quota_used_ratio: 100.00%
  arena_used_ratio: 40.5%
  items_used: 215912800
  quota_used: 536870912
  arena_size: 536870912
  arena_used: 217305440
...

localhost:3301> box.info.version
---
- 1.9.0-67-g3171288
...

localhost:3301> box.cfg
---
- log_nonblock: true
  checkpoint_count: 2
  rows_per_wal: 500000
  vinyl_range_size: 1073741824
  readahead: 16320
  log_level: 5
  memtx_min_tuple_size: 16
  vinyl_page_size: 8192
  coredump: false
  vinyl_run_size_ratio: 3.5
  replication_timeout: 1
  wal_max_size: 268435456
  memtx_memory: 536870912
  background: false
  vinyl_dir: .
  vinyl_cache: 134217728
  vinyl_read_threads: 1
  vinyl_timeout: 60
  worker_pool_threads: 4
  listen: '3301'
  too_long_threshold: 0.5
  vinyl_max_tuple_size: 1048576
  vinyl_write_threads: 2
  vinyl_bloom_fpr: 0.05
  log_format: plain
  memtx_max_tuple_size: 5242880
  memtx_dir: .
  wal_dir: .
  force_recovery: false
  wal_mode: write
  hot_standby: false
  vinyl_run_count_per_level: 2
  read_only: false
  replication_connect_timeout: 4
  replication_sync_lag: 10
  checkpoint_interval: 3600
  slab_alloc_factor: 1.05
  vinyl_memory: 134217728
  wal_dir_rescan_delay: 2
...
```

**Comment by locker** [comment]
I stumbled upon this issue during my investigation of the following test failure:
```
wal_off/lua.test.lua                                            [ fail ]

Test failed! Result content mismatch:
--- wal_off/lua.result  Mon Sep 18 13:55:15 2017
+++ wal_off/lua.reject  Tue Sep 25 19:44:39 2018
@@ -91,7 +91,7 @@
 ...
 mktuple(100000)
 ---
-- count 100000 len 368553
+- error: Failed to allocate 368569 bytes in slab allocator for memtx_tuple
 ...
 space:drop()
 ---

Reproduce file /home/vlad/src/tarantool/test/var/reproduce/001_wal_off.list.yaml
---
- [wal_off/tuple.test.lua, null]
- [wal_off/lua.test.lua, null]
...
```
Note, it occurs only if wal_off/lua runs after wal_off/tuple. The latter allocates tuples of increasing sizes, similarly to how it's described in this issue.

The allocation fails, because the small allocator, which is used for allocating tuples, maintains a memory pool for different allocation sizes, and a memory pool never frees the last slab, even if it's empty. According to the [comment](https://github.com/tarantool/small/blob/22d1bad1873edcb9b1383a273e70c4cf881d5349/small/mempool.h#L153) in the code, this is needed to prevent oscillations when a single object is repeatedly allocated and freed:

```C
struct mempool
{
	/**
	 * A completely empty slab which is not freed only to
	 * avoid the overhead of slab_cache oscillation around
	 * a single element allocation.
	 */
	struct mslab *spare;
};
```

We should probably limit the total amount of memory that can be used up by such spared slabs. May be, we could delete this logic altogether - after all, although it does prevent oscillations when there's a single object in a _mempool_, it doesn't help in case there's a single object in a _slab_ allocated for a mempool.

Here's a simple script that can be used to reliably reproduce this issue:
```Lua
box.cfg{
    memtx_memory = 512 * 1024 * 1024,
    memtx_max_tuple_size = 1024 * 1024
}

box.schema.space.create('test')
box.space.test:create_index('primary')

size = 1
while size < box.cfg.memtx_max_tuple_size - 128 do
    box.space.test:insert{1, string.rep('x', size)}
    box.space.test:delete{1}
    collectgarbage('collect')
    size = size * 1.05
end
```

**Comment by juise** [comment]
Stumbled upon the same issue on

```
> box.info.version
---
- 1.9.1-0-g06ec3d5
...
```

is there any ETA for bugfix?

**Comment by juise** [comment]
The possible workaround `slab_alloc_factor=2` suggested @kostja in https://groups.google.com/forum/#!topic/tarantool-ru/0EmWbax6g3w works very well

**Comment by nikolay-shmaleniuk** [comment]
`slab_alloc_factor` is deprecated

**Comment by zloidemon** [comment]
I've already tried play over options

```
    memtx_memory = 512 * 1024 * 1024,
    memtx_max_tuple_size = 1024 * 1024
```

But my issue solved after use `slab_alloc_factor=2` on `1.9.0-17-gc7c24f8`. tarantool works fine last 20 min.

**Comment by ZolotarevIgor-mail-ru** [comment]
We had seen this error recently in production. Cluster of about 20 storages, max storage size 30 GB, used 20 GB per storage and 4 replicators (instances that extract data from csv and load data to storages) 7/35 GB. Data stores in arrays on storages.
At first we had lost replication. Restart replicas solved the problem, but after about a month we had seen error "Failed to allocate bytes in slab allocator for memtx_tuple" and "failed to allocate bytes in mempool for new slab" on updating data in arrays on replicators, items_used_ratio > 90%, quota_used_ratio > 90%, arena_used_ratio = 20%. We had set slab_alloc_factor to 1.5 and error had been fixed.
